### ISSUE TITLE
How can I evaluate my model on a single class?

### ISSUE THREAD

**Issue by alexandre-eymael** [ultralytics/ultralytics#3952]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello!

I am using YOLOv8 for human detection. As I am only interested in the person class (i.e., class #0), I always call my model using `model.predict(image, classes=0)`.

I want to evaluate the performance of YOLOv8 on a custom dataset that only contains labeled humans. The dataset was created with Roboflow and works without any problems with the `model.val()` function. However, the `model.val()` function considers all classes when evaluating the model, while I am only interested in the detection of humans. This is what I am currently doing to obtain the desired information:

```python
metrics = model.val(dataset_path)
person_metrics = metrics.box.class_result(0) # Obtain metrics about the `Person` class (i.e., class 0)
```

But I am wondering whether the inclusion of all the other classes has an effect on the reported performance. When looking at the jpg files produced by the evaluation (see attached pictures), other objects are detected by the model, but these are totally irrelevant to me.

Therefore, I have two questions:

1) I am not sure I understand whether the presence of other classes negatively influences the model's performance. If there is no Potted Plant labeled in my dataset, but the model does predict a Potted Plant (correctly, but in my dataset, it is not labeled – I only labeled persons), does that influence the performance? (i.e., does it count as a false positive?) Or are the performances strictly computed class-wise (i.e., does the average precision of the "Person" class not depend on the presence or detection of other classes)?

2) Is there a way to only consider humans (i.e. the class #0) when calling `model.val()`? That would be ideal, as I would like the jpg files resulting from the `model.val()` call (see attached pictures) to only contain the "person" label (and not "Potted Plant", as it currently does).

Thank you!

Attached Files:
1) val_batch2_labels.jpg (ground truth):
![val_batch2_labels](https://github.com/ultralytics/ultralytics/assets/57938761/99907209-c1cf-426b-b52c-356b849e3944)

2) val_batch2_pred.jpg (prediction, contains other classes than persons, which is what I would like to remove)
![val_batch2_pred](https://github.com/ultralytics/ultralytics/assets/57938761/9e955843-1168-4e03-acdc-e13e3a6fe694)

**Comment by ammar-deep** [comment]
If you are evaluating a pretrained yolov8 model (provided by the authors) on your custom dataset you cannot evaluate it only on one class. 

You need to finetune the model on your custom dataset and specify the number of classes in the data.yaml file. Then when you evaluate this finetuned model it will only predict the class specified in the data.yaml file. 

An example of data.yaml file that you can use for finetuning the model on custom dataset is below.

```
path: /path/to/custom_dataset/parent_dir
train: images/train
val: images/val

names:
 0: person
nc: 1
```

I hope it helps.

**Comment by glenn-jocher** [comment]
@alexandre-eymael upon evaluation, YOLOv8 calculates performance metrics for each class separately and then averages them to obtain overall results. 

1) The performance alone on the `Person` class is independent of other classes. Even though the model might be predicting other classes (like `Potted Plant`), it doesn't affect the performance computation for the `Person` class. The presence of other classes influences overall performance, but not the individual class evaluation.

2) In terms of ignoring the detection of other classes and considering only humans, you could apply a post-processing step to filter out the detections belonging to other classes. However, this wouldn't affect the performance metrics as they are computed prior to this step. Currently, there isn't a direct way in `model.val()` to ignore detections of other classes. 

In your scenario, the metrics.box.class_result(0) is the appropriate method to observe the evaluation performance of the `Person` class. It will provide you with the Precision, Recall, and mAP values which are computed considering only the `Person` class. 

As for the presence of other class predictions in the visualization, it's simply because the model has been trained on datasets containing several other classes. If only humans are of interest, you might consider fine-tuning the model on only the `Person` class from your custom dataset.
